common/src/main/java/com/wynntils/core/json/JsonManager.java

```diff
@@ -66,19 +67,50 @@ public <T> T deepCopy(T value, Type fieldType) {
     public void savePreciousJson(File jsonFile, JsonObject jsonObject) {
         FileUtils.mkdir(jsonFile.getParentFile());
 
-        if (jsonFile.exists()) {
-            File backupFile = new File(jsonFile.getPath() + ".bak");
-            // Remove old backup (if any), and move current json file to backup
-            FileUtils.deleteFile(backupFile);
-            FileUtils.moveFile(jsonFile, backupFile);
+        File tempFile = new File(jsonFile.getPath() + ".tmp");
+
+        if (tempFile.exists()) {
+            // Clear an old temp file if it exists
+            FileUtils.deleteFile(tempFile);
         }
 
+        // Write the temp file
         try (OutputStreamWriter fileWriter =
-                new OutputStreamWriter(new FileOutputStream(jsonFile), StandardCharsets.UTF_8)) {
+                new OutputStreamWriter(new FileOutputStream(tempFile), StandardCharsets.UTF_8)) {
             GSON.toJson(jsonObject, fileWriter);
             fileWriter.flush();
         } catch (IOException e) {
-            WynntilsMod.error("Failed to save json file " + jsonFile, e);
+            WynntilsMod.error("Failed to save temp json file " + tempFile, e);
+        }
+
+        // Check that the temp file is valid before we overwrite the original file
+        try (FileReader reader = new FileReader(tempFile, StandardCharsets.UTF_8)) {
+            JsonElement parsed = JsonParser.parseReader(reader);
+
+            if (!parsed.isJsonObject()) {
+                WynntilsMod.error("Temporary json file " + tempFile + " did not contain a JsonObject");
+                return;
+            }
+        } catch (Exception e) {
+            WynntilsMod.error("Temporary json file " + tempFile + " is invalid", e);
+            return;
+        }
+
+        // Backup the old file, and replace it with the temp file
+        try {
+            if (jsonFile.exists()) {
+                File backupFile = new File(jsonFile.getPath() + ".bak");
+                // Remove old backup (if any), and move current json file to backup
+                FileUtils.deleteFile(backupFile);
+                FileUtils.moveFile(jsonFile, backupFile);
+            }
+
+            FileUtils.moveFile(tempFile, jsonFile);
+        } finally {
+            // Remove the temp file
+            if (tempFile.exists()) {
```

Comment:
If we get here, the temp file is corrupted somehow, or there is some serious problems with the file system. It might be worth keeping it for analysis sake. Also, should we try to save it again you will start by removing the temp file.